ic-os/components/networking/setup-networking/setup-networking.sh

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env bash
+
+set -o nounset
+set -o pipefail
+
+SHELL="/bin/bash"
+PATH="${PATH}:/sbin:/bin:/usr/sbin:/usr/bin"
+
+function parse_args() {
+    if [ $# -ne 1 ]; then
+        echo "Usage: $0 <SetupOS|HostOS>"
+        exit 1
+    fi
+
+    NODE_TYPE="$1"
+    if [ "$NODE_TYPE" != "SetupOS" ] && [ "$NODE_TYPE" != "HostOS" ]; then
+        echo "Invalid node type: $NODE_TYPE"
+        exit 1
+    fi
+
+    # Allow overriding config paths via environment variables
+    if [ "$NODE_TYPE" = "SetupOS" ]; then
+        CONFIG_BASE_PATH="${CONFIG_BASE_PATH:-/var/ic/config}"
+    else
+        CONFIG_BASE_PATH="${CONFIG_BASE_PATH:-/boot/config}"
+    fi
+
+    CONFIG="${CONFIG_BASE_PATH}/config.ini"
+}
+
+function read_variables() {
+    # Read limited set of keys. Be extra-careful quoting values as it could
+    # otherwise lead to executing arbitrary shell code!
+    while IFS="=" read -r key value; do
+        case "$key" in
+            "ipv6_gateway") IPV6_GATEWAY="${value}" ;;
+        esac
+    done <"${CONFIG}"
+
+    if [ -z "$IPV6_GATEWAY" ]; then
+        echo "No IPv6 gateway found in $CONFIG."
+        exit 1
+    fi
+}
+
+function generate_addresses() {
+    echo "Generating MAC and IPv6 addresses..."
+    IC_BIN_PATH="${IC_BIN_PATH:-/opt/ic/bin}"
+
+    if [ "$NODE_TYPE" = "SetupOS" ]; then
+        MAC_ADDR=$("${IC_BIN_PATH}/setupos_tool" generate-mac-address --node-type SetupOS)
+        IPV6_ADDR=$("${IC_BIN_PATH}/setupos_tool" generate-ipv6-address --node-type SetupOS)
+    else
+        MAC_ADDR=$("${IC_BIN_PATH}/hostos_tool" generate-mac-address --node-type HostOS)
+        IPV6_ADDR=$("${IC_BIN_PATH}/hostos_tool" generate-ipv6-address --node-type HostOS)
+    fi
+
+    if [ -z "$MAC_ADDR" ] || [ -z "$IPV6_ADDR" ]; then
+        echo "Failed to generate MAC or IPv6 address."
+        exit 1
+    fi
+
+    echo "Generated MAC address: $MAC_ADDR"
+    echo "Generated IPv6 address: $IPV6_ADDR"
+}
+
+function gather_interfaces_by_speed() {
+    echo "Gathering and sorting interfaces by speed..."
+
+    SYS_NET_PATH="${SYS_NET_PATH:-/sys/class/net}"
+    INTERFACES=($(find "$SYS_NET_PATH" -type l -not -lname '*virtual*' -exec basename '{}' ';'))
+    declare -A SPEED_MAP
+
+    for IFACE in "${INTERFACES[@]}"; do
+        if [ -f "$SYS_NET_PATH/$IFACE/speed" ]; then
+            SPEED=$(cat "$SYS_NET_PATH/$IFACE/speed" 2>/dev/null || echo 0)
```

Comment:
I want to draw attention to the way we're gathering interface speeds. I'm curious what PFOPS thinks of this. Does this seem proper? In the old code, we were parsing the output of Ethtool to gather speeds (see interfaces.rs)